python/ray/ml/train/integrations/huggingface/huggingface_utils.py

```diff
@@ -0,0 +1,203 @@
+
+import datasets.iterable_dataset
+import transformers.trainer
+from torch.utils.data import IterableDataset, DataLoader
+from transformers.trainer_callback import TrainerCallback
+
+from ray import train
+from ray.util import get_node_ip_address
+from ray.data.dataset import Dataset
+
+# Constants for the sync checkpoint dict. See huggingface_trainer.py
+CHECKPOINT_PATH_ON_NODE_KEY = "checkpoint_path_on_node"
+NODE_IP_KEY = "node_ip"
+
+
+def maybe_add_length(obj: Any, length: Optional[int]) -> Any:
+    """Change the class of obj to a subclass with predefined __len__ if needed."""
+    # By adding length to the dataset we let HF calculate steps per epoch
+    # and other such values. Without length, it's not possible to use
+    # epochs as the evaluation strategy, which makes for poor UX.
+
+    if not length or hasattr(obj, "__len__"):
+        return obj
+
+    def __len__(self):
+        return length
+
+    new_class = type(
+        f"{obj.__class__.__name__}WithLength", (obj.__class__,), {"__len__": __len__}
+    )
+    obj.__class__ = new_class
+    return obj
+
+
+def wrap_transformers_trainer(
+    trainer: transformers.trainer.Trainer,
+) -> transformers.trainer.Trainer:
+    """Change the class of trainer to a subclass implementing Ray-specific logic."""
+    base_trainer_class: Type[transformers.trainer.Trainer] = trainer.__class__
+
+    class RayTrainer(base_trainer_class):
+        # TODO(yard1): Upstream data collator removing unused columns to
+        # transformers.
+        # This is necessary to provide the same experience as with a
+        # non-iterable HuggingFace Dataset, which can remove columns
+        # not supported by the model.
+        def _prepare_data_collator(self):
+            """Wrap the data collator in a function removing superflous columns."""
+            # Hack to set the self._signature_columns attribute.
+            try:
+                self._remove_unused_columns(None, description="nan")
+            except AttributeError:
+                pass
+
+            if self._signature_columns and not hasattr(self, "_original_data_collator"):
+                self._original_data_collator = self.data_collator
+
+                def remove_columns_collator(features):
+                    features = [
+                        {
+                            k: v
+                            for k, v in feature.items()
+                            if k in self._signature_columns
+                        }
+                        for feature in features
+                    ]
+                    return self._original_data_collator(features)
+
+                collator = remove_columns_collator
+            else:
+                collator = self.data_collator
+
+            self.data_collator = collator
+
+        def get_train_dataloader(self):
+            if self.train_dataset is None:
+                raise ValueError("Trainer: training requires a train_dataset.")
```

Comment:
Is this error message exposed to users? If so, can we have a better message here?

Reply:
This is the same message & validation as in HuggingFace.